libobs/obs-cocoa.m

```diff
@@ -187,702 +434,392 @@ static bool dstr_from_cfstring(struct dstr *str, CFStringRef ref)
+    const UniCharCount maxLength = 16;
+    UniChar buffer[16];
 
-    input_method_changed(NULL, hotkeys, NULL, NULL, NULL);
-    return hotkeys->platform_context != NULL;
-}
+    if (obs_key_to_localized_string(key, str)) {
+        return;
+    }
 
-void obs_hotkeys_platform_free(struct obs_core_hotkeys *hotkeys)
-{
-    CFNotificationCenterRemoveEveryObserver(CFNotificationCenterGetDistributedCenter(), hotkeys);
+    int code = obs_key_to_virtual_key(key);
+    if (key_code_to_string(code, str)) {
+        return;
+    }
 
-    hotkeys_release(hotkeys->platform_context);
-}
+    if (code == INVALID_KEY) {
+        const char *keyName = obs_key_to_name(key);
+        blog(LOG_ERROR, "hotkey-cocoa: Got invalid key while translating '%d' (%s)", key, keyName);
+        dstr_copy(str, keyName);
+        return;
+    }
 
-typedef unsigned long NSUInteger;
+    struct obs_hotkeys_platform *platform = NULL;
 
-static bool mouse_button_pressed(obs_key_t key, bool *pressed)
-{
-    int button = 0;
-    switch (key) {
-#define MAP_BUTTON(n)      \
-    case OBS_KEY_MOUSE##n: \
-        button = n - 1;    \
-        break
-        MAP_BUTTON(1);
-        MAP_BUTTON(2);
-        MAP_BUTTON(3);
-        MAP_BUTTON(4);
-        MAP_BUTTON(5);
-        MAP_BUTTON(6);
-        MAP_BUTTON(7);
-        MAP_BUTTON(8);
-        MAP_BUTTON(9);
-        MAP_BUTTON(10);
-        MAP_BUTTON(11);
-        MAP_BUTTON(12);
-        MAP_BUTTON(13);
-        MAP_BUTTON(14);
-        MAP_BUTTON(15);
-        MAP_BUTTON(16);
-        MAP_BUTTON(17);
-        MAP_BUTTON(18);
-        MAP_BUTTON(19);
-        MAP_BUTTON(20);
-        MAP_BUTTON(21);
-        MAP_BUTTON(22);
-        MAP_BUTTON(23);
-        MAP_BUTTON(24);
-        MAP_BUTTON(25);
-        MAP_BUTTON(26);
-        MAP_BUTTON(27);
-        MAP_BUTTON(28);
-        MAP_BUTTON(29);
-        break;
-#undef MAP_BUTTON
-
-        default:
-            return false;
+    if (obs) {
+        pthread_mutex_lock(&obs->hotkeys.mutex);
+        platform = obs->hotkeys.platform_context;
+        hotkeys_retain(platform);
+        pthread_mutex_unlock(&obs->hotkeys.mutex);
     }
 
-    NSUInteger buttons = [NSEvent pressedMouseButtons];
-    *pressed = (buttons & (1 << button)) != 0;
-    return true;
-}
+    if (!platform) {
+        const char *keyName = obs_key_to_name(key);
+        blog(LOG_ERROR, "hotkey-cocoa: Could not get hotkey platform while translating '%d' (%s)", key, keyName);
+        dstr_copy(str, keyName);
+        return;
+    }
 
-bool obs_hotkeys_platform_is_pressed(obs_hotkeys_platform_t *plat, obs_key_t key)
-{
-    bool mouse_pressed = false;
-    if (mouse_button_pressed(key, &mouse_pressed))
-        return mouse_pressed;
+    UInt32 deadKeyState = 0;
+    UniCharCount length = 0;
 
-    if (!plat)
-        return false;
+    OSStatus err = UCKeyTranslate(platform->layout, code, kUCKeyActionDown, 0x104, LMGetKbdType(),
```

Comment:
We lost the comment here about 0x104 being related to caps lock, which may have been useful/informative for reading the code. Maybe worth defining 0x104 as a constant?

Reply:
Per the docs [here](https://developer.apple.com/documentation/coreservices/1390584-uckeytranslate), this is derived from some sort of bit shift from a Carbon modifier key code masked with `0xFF`, but I can't figure out the exact derivation for caps lock in particular...

Reply:
Maybe a stupid question, but how can _any_ value logically combined with `0xFF` ever produce a value larger than `0xFF` (as `0x104` is)?

That value seems invalid to me, because the only value that remains from the bit mask is bit 3 set to `1` (`0x40`).

The `modifiers` member of the `EventRecord` struct is a `UInt16` (aka `EventModifiers`), and caps lock (aka alphaLock) is bit 10 (`0x400`) and if I shift that value by 8 I get `0x4`.

Reply:
Funnily enough what might happen here is that the 9th bit might just be ignored by macOS and it "sees" just the `0x04` part, which would align again.

Reply:
I presume that must be the case; it looks like all of the other Carbon modifier key enums would also have ignored bits when shifted (though they do all end up being unique at least); I have no idea why this would be done this way.

It looks like the purpose of this block of code in general is to ensure that the string representation of the chosen hotkey is always capitalized, I guess so OBS doesn't imply that capitalization state matters for letter-key keybinds.

I confirmed that just putting in `4` works to achieve the same effect.

It seems like we definitely wouldn't want to import Carbon just for the sake of having the actual `alphaLock` available for defining this goofy constant, so maybe we just do something like `UInt32 carbonCapsLockBitMask = 0x4` and provide `carbonCapsLockBitMask` to `UCKeyTranslate` for readability?

Reply:
We already include the Carbon header anyway for the other key codes, so we can use the "official" values.

Reply:
I use `alphaLock` now and downshift + mask it as expected by the documentation for `UCKeyTranslate`.